src/main/java/com/github/kay9/dragonmounts/dragon/ai/DragonFollowOwnerGoal.java

```diff
@@ -0,0 +1,179 @@
+package com.github.kay9.dragonmounts.dragon.ai;
+
+import com.github.kay9.dragonmounts.dragon.TameableDragon;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.ai.goal.Goal;
+import net.minecraft.world.entity.ai.navigation.FlyingPathNavigation;
+import net.minecraft.world.entity.ai.navigation.GroundPathNavigation;
+import net.minecraft.world.level.LevelReader;
+import net.minecraft.world.level.block.LeavesBlock;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.pathfinder.BlockPathTypes;
+import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
+
+import java.util.EnumSet;
+
+/**
+ * Goal for dragon to follow its owner.
+ * <p></p>
+ * Mostly copied from <code>FollowOwnerGoal</code>, but with some modifications to fix an issue.
+ * Also allows dragon to tp to owner in the air so they don't get stuck until the owner lands.
+ *
+ * @author AnimalsWritingCode
+ *
+ * @see net.minecraft.world.entity.ai.goal.FollowOwnerGoal
+ */
+public class DragonFollowOwnerGoal extends Goal {
+    private static final int MIN_HORIZONTAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 2;
+    private static final int MAX_HORIZONTAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 3;
+    private static final int MIN_VERTICAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 0;
+    private static final int MAX_VERTICAL_DISTANCE_FROM_PLAYER_WHEN_TELEPORTING = 1;
+    private final TameableDragon dragon;
+    private final LevelReader level;
+    private final double speedModifier;
+    private int timeToRecalcPath;
+    private final float stopDistance;
+    private final float startDistance;
+    private final float teleportDistance;
+    private float oldWaterCost;
+
+    public DragonFollowOwnerGoal(TameableDragon dragon, double speedModifier, float startDistance, float stopDistance, float teleportDistance) {
+        this.dragon = dragon;
+        this.level = dragon.level;
+        this.speedModifier = speedModifier;
+        this.startDistance = startDistance;
+        this.stopDistance = stopDistance;
+        this.teleportDistance = teleportDistance;
+        this.setFlags(EnumSet.of(Goal.Flag.MOVE, Goal.Flag.LOOK));
+        if (!(dragon.getNavigation() instanceof GroundPathNavigation) && !(dragon.getNavigation() instanceof FlyingPathNavigation)) {
```

Comment:
This check seems redundant as under normal circumstances the dragons never have a different navigator.

Reply:
I actually left this here in case some kind of swimming navigation was done, but yeah I guess that feels unnecessary. Axing it.

Reply:
ah, I see your reasoning now.
My plan for the swimming navigator will come with the ai overhaul anyway, which would also include scrapping the goal system, for now I feel this is best.
Thanks!